sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/signaling/CommunicationSignalingClient.java

```diff
@@ -79,20 +97,37 @@ public String getSkypetoken(boolean forceRefresh) {
     /**
      * Start the realtime connection.
      * @param context the android application context
+     * @param errorHandler error handler callback for registration failures
      */
-    public void start(Context context) {
+    public void start(Context context, Consumer<Throwable> errorHandler) {
         ISkypetokenProvider skypetokenProvider = new ISkypetokenProvider() {
             @Override
             public String getSkypetoken(boolean forceRefresh) {
+                if (forceRefresh) {
+                    tokenFetchRetries += 1;
+                    if (tokenFetchRetries > NotificationUtils.MAX_TOKEN_FETCH_RETRY_COUNT) {
+                        stop();
+                        Throwable throwable = new Throwable("Access token is expired and failed to fetch a valid one after "
+                            + NotificationUtils.MAX_TOKEN_FETCH_RETRY_COUNT
+                            + " retries.");
+                        logger.logThrowableAsError(throwable);
+                        errorHandler.accept(throwable);
+                        return null;
+                    }
+                } else {
+                    tokenFetchRetries = 0;
+                }
+
+                String skypeUserToken = null;
                 try {
-                    return communicationTokenCredential.getToken().get().getToken();
-                } catch (InterruptedException e) {
-                    throw logger.logExceptionAsError(
-                        new RuntimeException("Get skype user token failed for realtime notification: " + e.getMessage()));
-                } catch (ExecutionException e) {
-                    throw logger.logExceptionAsError(
-                        new RuntimeException("Get skype user token failed for realtime notification: " + e.getMessage()));
+                    skypeUserToken = communicationTokenCredential.getToken().get().getToken();
+                } catch (InterruptedException | ExecutionException e) {
+                    logger.error("Get skype user token failed for realtime notification: " + e.getMessage());
+                    // Return a empty but not null skype user token to trigger retry
+                    skypeUserToken = "";
```

Comment:
Where is this retry with an empty token triggered? I thought the `errorHandler` would be the mechanism a customer could use to retry this operation.

Reply:
If this `getSkypetoken` callback does not return a valid Skypetoken, trouter library would always retry to call this function with passing `forceRefresh = true`. So, if we find `forceRefresh == true` (line 106) and exceed retry limit (line 108), it means we are not able to get a valid Skypetoken after retry. That time we would call `errorHandler` to let customer know, we failed to start realtime notification and they may need to do something else. In that customer implemented `errorHandler`, they may update credential, etc.

Reply:
In short words, "Where is this retry with an empty token triggered?" - inside trouter lib.